dnf5/commands/check/check.cpp

```diff
@@ -0,0 +1,323 @@
+                std::string_view problem_str = problem_cstr;
+                auto cnevr = rpmProblemGetPkgNEVR(p);
+                libdnf_assert(cnevr != NULL, "command \"check\": rpmProblemGetPkgNEVR() returns NULL");
+                std::string_view nevr = cnevr;
+                auto alt_cnevr = rpmProblemGetAltNEVR(p);
+                libdnf_assert(alt_cnevr != NULL, "command \"check\": rpmProblemGetAltNEVR() returns NULL");
+                std::string_view alt_nevr = alt_cnevr;
+
+                switch (rpmProblemGetType(p)) {
+                    case RPMPROB_REQUIRES:
+                        if (all || dependencies) {
+                            problems[std::string{alt_nevr}].insert(Problem{
+                                .type = ProblemType::MISSING_REQUIRE,
+                                .nevra = "",
+                                .file_or_provide = std::string{problem_str}});
+                        }
+                        break;
+                    case RPMPROB_CONFLICT:
+                        if (all || dependencies) {
+                            auto nevr_compare = nevr.compare(alt_nevr);
+                            if (nevr_compare == 0) {
+                                // skip self conflicts
+                                break;
+                            }
+                            problems[std::string{nevr}].insert(Problem{
+                                .type = ProblemType::CONFLICT,
+                                .nevra = std::string{alt_nevr},
+                                .file_or_provide = std::string{problem_str}});
+                        }
+                        break;
+                    case RPMPROB_OBSOLETES:
+                        if (all || obsoleted) {
+                            auto nevr_compare = nevr.compare(alt_nevr);
+                            if (nevr_compare == 0) {
+                                // skip self obsolete
+                                break;
+                            }
+                            problems[std::string{nevr}].insert(Problem{
+                                .type = ProblemType::OBSOLETED,
+                                .nevra = std::string{alt_nevr},
+                                .file_or_provide = std::string{problem_str}});
+                        }
+                        break;
+                    default:;
+                }
+            }
+            rpmpsFreeIterator(psi);
+        }
+        rpmpsFree(ps);
+        rpmtsEmpty(ts);
+    }
+    return rc;
+}
+
+
+// Groups packages with the same "name" and "architecture".
+// For "installonly" packages, only one "nevra" is inserted.
+int group_pkgs_same_name_arch([[maybe_unused]] QVA_t qva, [[maybe_unused]] rpmts ts, Header h) {
+    auto * arch = headerGetString(h, RPMTAG_ARCH);
+    if (!arch) {
+        // skip special package representing key installed in the rpm database
```

Comment:
Interesting way of detecting gpg keys. Libsolv uses `memcmp(key, "gpg-pubkey", 10)` for the same purpose. @pmatilai Is there any chance to handle GPG keys differently in RPM? Or is there any smart or recommended detection in RPM whether element is package or a key?

Reply:
A more accurate comment would be: "skip any non-packages (such as gpg-pubkey) in the database"